### PR TITLE
Tokotron: Tokenized TTS for the SpeechBrain benchmark (single speaker)

### PR DESCRIPTION
**Pull request by flexthink**

### CODE REVIEW

File: benchmarks/DASB/utils/eval.py
[the remark this answers is not in the document]
Comment (flexthink) [reply benchmarks/DASB/utils/eval.py:1]: A separate eval is more flexible in running multiple evaluations against a single training without modifying anything in the folder, particularly for experiments that take a long time to run. It is more relevant for the main Tokotron branch than for the benchmark.

File: benchmarks/DASB/utils/train_logger.py
[the remark this answers is not in the document]
Comment (poonehmousavi) [reply benchmarks/DASB/utils/train_logger.py:1]: for benchmark, it is not necessary since people not necessary use SLURM

File: benchmarks/DASB/utils/eval.py
[the remark this answers is not in the document]
Comment (poonehmousavi) [reply benchmarks/DASB/utils/eval.py:1]: for the benchmark we don't need that flexibility, let's fix a set of metrics and always report them.,, in that case, we could have one py file that do both training nd eval

File: benchmarks/DASB/utils/__init__.py
[the remark this answers is not in the document]
Comment (flexthink) [reply benchmarks/DASB/utils/__init__.py:1]: The need for utilities is on a case by case basis - most of these are because a benchmark PR does not make it possible to enhance the core.

`data`: Fixes a performance issue in the core `undo_batch` implementation, which insists on converting tensors to lists. There is also a function

`preparation`: Avoids having to run feature extraction at the beginning of every experiment or inline - it helps convert waves to token tensors during preparation (similar to how existing recipes run G2P and other expensive preparation steps) - and then packs it as a tarball, avoiding performance and quota issues with SLURM. Reusable for different datasets, not specific to TTS.

`eval`: Reusable and extensible ASR and MOS evaluation, avoids having to copy and paste them into multiple recipes. This is shared between single-speaker and multi-speaker.


`hparams`: Needed to be able to override the list of layers with a command-line argument, which is not natively supported in hyperpyyaml (you can't override lists - only strings or scalars)... This is specific to the Benchmark. A cleaner alternative is allowing DiscreteSSL to accept lists of layers as a comma-separated string. `repeat_for_layers` works around hard-coded layer lists in defaults (a cleaner alternative is to default to [None] * layer_count instead)

`tts`: Saving samples, plotting attention, plotting gate activation diagrams. This is shared between single-speaker and multi-speaker.

`audio_tokens`: Computes silence padding, not currently used outside of TTS - but usable for any task where silence padding is preferable to zero padding.

[the remark this answers is not in the document]
Comment (poonehmousavi) [reply benchmarks/DASB/utils/__init__.py:1]: could we move some of these utils to Speechbrain?

File: benchmarks/DASB/utils/audio_tokens.py
[the remark this answers is not in the document]
Comment (flexthink) [reply benchmarks/DASB/utils/audio_tokens.py:1]: I can move them to tokotron or custom_model

**Comment by poonehmousavi** [comment]
@flexthink  any update on this PR?

File: benchmarks/DASB/model/custom_model.py [hunk 1]
```diff
@@ -94,3 +106,461 @@ def forward(self, in_tokens):
...
+        emb : torch.Tensor
+            a (Batch x Length x Heads x Embedding) tensor
+            of raw vector embeddings from the model's
+            quantizer codebooks
+        """
+        emb, _, _ = self.dac.quantizer.from_codes(tokens.transpose(1, 2).int())
+        return emb.transpose(1, 2)
+
+
+class SpeechTokenizerInterface(nn.Module):
```
[the remark this answers is not in the document]
Comment (flexthink) [reply benchmarks/DASB/model/custom_model.py:278]: It was not available in the version of SpeechBrain being used.

**Review by poonehmousavi — CHANGES_REQUESTED** [review]
since util is only used in TTS.. it would be better to move it to TTS folder

**Review by poonehmousavi — CHANGES_REQUESTED** [review]
the wrappers added for all tokenzier  are only used in TTS.. so maybe better to cerate the custom_model specific to TTS so it is less confusing.. but the diea is really good . in the next refactoring phase, I will try to apply teh changes across all tokenizers

File: benchmarks/DASB/model/custom_model.py [hunk 2]
```diff
@@ -1,4 +1,15 @@
 import torch
+from torch import nn
+from speechbrain.dataio.dataio import length_to_mask
+import math
+import logging
+
+from huggingface_hub import snapshot_download
+
+try:
+    from speechtokenizer import SpeechTokenizer
```
Comment (poonehmousavi) [inline benchmarks/DASB/model/custom_model.py:10]: precommit error for unused lib

**Comment by poonehmousavi** [comment]
Thanks @flexthink, everything looks really neat .. I  did the review... I have fetched the latest update from the DASB main branch... These are the few comments that I have before merging the PR:

1. There is some recommit error mostly related to unused lib.
2. The utils and Custim_model should be moved to TTS folder, especially custom_model should be renamed or avoid the link to the main custim_model because the wrappers ae not used for other tasks and it could be confusing,.. we could add your changes later to all tasks, but for now let's keep it only for TTS.
3. I have added the run_generative_benchmark.sh for SE and SS, please complete it for TTS and make sure that it works.

File: benchmarks/DASB/model/custom_model.py (hunk 2 above)
Comment (flexthink) [reply benchmarks/DASB/model/custom_model.py:10]: No longer needed given some of the code has been moved out, will update

**Comment by poonehmousavi** [comment]
Thanks @flexthink  for this PR... everything works now..It merged
